File: validation/known_assignment_mml_crossfit.R
#!/usr/bin/env Rscript

# Independent base-R evaluator/refitter for the registered free-SD PCM MML lane.
# It intentionally does not source Python or use a Python-derived R package.

parse_args <- function(values) {
  result <- list()
  index <- 1L
  while (index <= length(values)) {
    key <- sub("^--", "", values[[index]])
    if (index == length(values)) stop("Missing value for --", key, call. = FALSE)
    result[[key]] <- values[[index + 1L]]
    index <- index + 2L
  }
  result
}

`%||%` <- function(value, replacement) {
  if (is.null(value) || !length(value)) replacement else value
}

args <- parse_args(commandArgs(trailingOnly = TRUE))
required <- c("study", "output", "vectors")
missing <- required[!vapply(required, function(key) nzchar(args[[key]] %||% ""), logical(1))]
if (length(missing)) stop("Missing arguments: ", paste(missing, collapse = ", "), call. = FALSE)

study_dir <- normalizePath(args$study, mustWork = TRUE)
output_dir <- normalizePath(args$output, mustWork = FALSE)
dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)
vector_ids <- as.integer(strsplit(args$vectors, ",", fixed = TRUE)[[1L]])
if (anyNA(vector_ids) || !length(vector_ids)) stop("--vectors must contain integers", call. = FALSE)
maxit <- as.integer(args$maxit %||% "400")
if (!is.finite(maxit) || maxit < 1L) stop("--maxit must be positive", call. = FALSE)

input_dir <- file.path(study_dir, "retained_input")
manifest <- utils::read.csv(file.path(input_dir, "manifest.csv"), stringsAsFactors = FALSE)
ratings_all <- utils::read.csv(file.path(input_dir, "generated_ratings.csv"), stringsAsFactors = FALSE)
attempts <- utils::read.csv(file.path(input_dir, "attempt_manifest.csv"), stringsAsFactors = FALSE)
attempts <- attempts[
  attempts$AttemptType == "PYTHON_MML_FREE_SD_Q31_PCM" &
    attempts$PersonVector %in% vector_ids,
  , drop = FALSE
]
if (nrow(attempts) != 3L * length(unique(vector_ids))) {
  stop("Expected three free-MML attempts per selected Person vector", call. = FALSE)
}

logsumexp <- function(values) {
  maximum <- max(values)
  maximum + log(sum(exp(values - maximum)))
}

normal_quadrature <- function(n, sigma) {
  # Golub-Welsch construction for physicists' Hermite quadrature.  The first
  # eigenvector row squared is already normalized to N(0,1) weights after the
  # sqrt(2) node transformation.
  off <- sqrt((1L:(n - 1L)) / 2)
  jacobi <- matrix(0, nrow = n, ncol = n)
  jacobi[cbind(1L:(n - 1L), 2L:n)] <- off
  jacobi[cbind(2L:n, 1L:(n - 1L))] <- off
  eig <- eigen(jacobi, symmetric = TRUE)
  order_index <- order(eig$values)
  list(
    nodes = sqrt(2) * eig$values[order_index] * sigma,
    weights = eig$vectors[1L, order_index]^2
  )
}

expand_parameters <- function(par) {
  if (length(par) != 12L) stop("Expected 12 coordinates", call. = FALSE)
  list(
    Rater = c(par[1:3], -sum(par[1:3])),
    Task = c(par[4:5], -sum(par[4:5])),
    Criterion = par[6:7],
    steps = rbind(
      C01 = c(par[8:9], -sum(par[8:9])),
      C02 = c(par[10:11], -sum(par[10:11]))
    ),
    sigma = exp(par[12])
  )
}

prepare_data <- function(data) {
  data$Person <- factor(data$Person, levels = sort(unique(data$Person)))
  data$Rater <- factor(data$Rater, levels = c("R01", "R02", "R03", "R04"))
  data$Task <- factor(data$Task, levels = c("T01", "T02", "T03"))
  data$Criterion <- factor(data$Criterion, levels = c("C01", "C02"))
  data$Score <- as.integer(data$Score)
  if (anyNA(data[, c("Person", "Rater", "Task", "Criterion", "Score")])) {
    stop("Cross-fit data contains unmapped values", call. = FALSE)
  }
  list(
    data = data,
    person = as.integer(data$Person),
    rater = as.integer(data$Rater),
    task = as.integer(data$Task),
    criterion = as.integer(data$Criterion),
    score = data$Score,
    n_person = nlevels(data$Person)
  )
}

marginal_components <- function(par, prepared, q = 31L, retain_posterior = FALSE) {
  expanded <- expand_parameters(par)
  quadrature <- normal_quadrature(q, expanded$sigma)
  log_weights <- log(quadrature$weights)
  base_eta <- -expanded$Rater[prepared$rater] -
    expanded$Task[prepared$task] -
    expanded$Criterion[prepared$criterion]
  cumulative <- t(apply(expanded$steps, 1L, function(value) c(0, cumsum(value))))
  threshold_by_row <- cumulative[prepared$criterion, , drop = FALSE]
  categories <- 0:3
  person_node_ll <- matrix(0, nrow = prepared$n_person, ncol = q)
  row_index <- seq_along(prepared$score)
  for (node_index in seq_len(q)) {
    eta <- quadrature$nodes[[node_index]] + base_eta
    logits <- outer(eta, categories) - threshold_by_row
    maximum <- pmax(logits[, 1L], logits[, 2L], logits[, 3L], logits[, 4L])
    denominator <- maximum + log(rowSums(exp(logits - maximum)))
    observed <- logits[cbind(row_index, prepared$score + 1L)]
    by_person <- rowsum(observed - denominator, prepared$person, reorder = FALSE)
    person_node_ll[, node_index] <- by_person[, 1L]
  }
  joint <- sweep(person_node_ll, 2L, log_weights, "+")
  maximum <- apply(joint, 1L, max)
  person_values <- maximum + log(rowSums(exp(joint - maximum)))
  posterior <- if (retain_posterior) exp(joint - person_values) else NULL
  list(
    loglik = sum(person_values),
    posterior = posterior,
    nodes = quadrature$nodes,
    weights = quadrature$weights,
    expanded = expanded
  )
}

objective <- function(par, prepared, q = 31L) {
  value <- tryCatch(-marginal_components(par, prepared, q = q)$loglik, error = function(error) Inf)
  if (is.finite(value)) value else .Machine$double.xmax / 100
}

finite_difference_gradient <- function(par, prepared, q = 31L, step = 1e-5) {
  gradient <- numeric(length(par))
  for (index in seq_along(par)) {
    local_step <- step * max(1, abs(par[[index]]))
    plus <- minus <- par
    plus[[index]] <- plus[[index]] + local_step
    minus[[index]] <- minus[[index]] - local_step
    gradient[[index]] <- (objective(plus, prepared, q) - objective(minus, prepared, q)) /
      (2 * local_step)
  }
  gradient
}

python_start <- function(recovery, thresholds, run) {
  facet <- function(name, levels) {
    rows <- recovery[recovery$Facet == name, , drop = FALSE]
    values <- rows$Estimate[match(levels, rows$Level)]
    if (anyNA(values)) stop("Missing Python facet coordinate: ", name, call. = FALSE)
    as.numeric(values)
  }
  rater <- facet("Rater", c("R01", "R02", "R03", "R04"))
  task <- facet("Task", c("T01", "T02", "T03"))
  criterion <- facet("Criterion", c("C01", "C02"))
  steps <- lapply(c("C01", "C02"), function(level) {
    rows <- thresholds[thresholds$StepFacetLevel == level, , drop = FALSE]
    rows <- rows[order(rows$Category), , drop = FALSE]
    as.numeric(rows$Estimate)
  })
  if (any(vapply(steps, length, integer(1)) != 3L)) stop("Missing Python PCM steps", call. = FALSE)
  sigma <- as.numeric(run$EstimatedPopulationSD[[1L]])
  if (!is.finite(sigma) || sigma <= 0) stop("Invalid Python population SD", call. = FALSE)
  c(rater[1:3], task[1:2], criterion, steps[[1L]][1:2], steps[[2L]][1:2], log(sigma))
}

run_rows <- list()
parameter_rows <- list()
gradient_rows <- list()
run_index <- 0L
parameter_index <- 0L
gradient_index <- 0L
coordinate_names <- c(
  "Rater::R01", "Rater::R02", "Rater::R03",
  "Task::T01", "Task::T02", "Criterion::C01", "Criterion::C02",
  "Step::C01::1", "Step::C01::2", "Step::C02::1", "Step::C02::2",
  "LogPopulationSD"
)

for (attempt_index in seq_len(nrow(attempts))) {
  attempt <- attempts[attempt_index, , drop = FALSE]
  run_id <- as.character(attempt$RunId[[1L]])
  ordinal <- as.integer(attempt$AttemptOrdinal[[1L]])
  artifact <- file.path(study_dir, "work", sprintf("%05d", ordinal))
  run <- utils::read.csv(file.path(artifact, "run_ledger.csv"), stringsAsFactors = FALSE)
  recovery <- utils::read.csv(file.path(artifact, "recovery.csv"), stringsAsFactors = FALSE)
  thresholds <- utils::read.csv(file.path(artifact, "thresholds.csv"), stringsAsFactors = FALSE)
  data <- ratings_all[ratings_all$RunId == run_id, c("Person", "Rater", "Task", "Criterion", "Score")]
  prepared <- prepare_data(data)
  start <- python_start(recovery, thresholds, run)
  python_q31 <- marginal_components(start, prepared, q = 31L, retain_posterior = TRUE)
  python_q61 <- marginal_components(start, prepared, q = 61L)
  python_gradient <- finite_difference_gradient(start, prepared, q = 31L)
  sigma_fixed_point <- sqrt(sum(python_q31$posterior * rep(python_q31$nodes^2, each = nrow(python_q31$posterior))) /
    nrow(python_q31$posterior))

  fit <- optim(
    par = start,
    fn = objective,
    gr = finite_difference_gradient,
    prepared = prepared,
    q = 31L,
    method = "BFGS",
    control = list(maxit = maxit, reltol = 1e-12)
  )
  fit_gradient <- finite_difference_gradient(fit$par, prepared, q = 31L)
  fit_q61 <- marginal_components(fit$par, prepared, q = 61L)
  fit61 <- optim(
    par = fit$par,
    fn = objective,
    gr = finite_difference_gradient,
    prepared = prepared,
    q = 61L,
    method = "BFGS",
    control = list(maxit = maxit, reltol = 1e-12)
  )
  fit61_gradient <- finite_difference_gradient(fit61$par, prepared, q = 61L)
  expanded_python <- expand_parameters(start)
  expanded_r <- expand_parameters(fit$par)
  expanded_r61 <- expand_parameters(fit61$par)

  run_index <- run_index + 1L
  run_rows[[run_index]] <- data.frame(
    RunId = run_id,
    PersonVector = as.integer(attempt$PersonVector[[1L]]),
    Gamma = as.numeric(attempt$Gamma[[1L]]),
    PythonRecordedLogLik = as.numeric(run$LogLik[[1L]]),
    PythonSolutionRLogLikQ31 = python_q31$loglik,
    PythonSolutionRLogLikQ61 = python_q61$loglik,
    PythonSolutionRGradientSupNormQ31 = max(abs(python_gradient)),
    PythonSigma = expanded_python$sigma,
    PythonSigmaFixedPoint = sigma_fixed_point,
    PythonSigmaFixedPointResidual = sigma_fixed_point - expanded_python$sigma,
    RConvergenceCode = as.integer(fit$convergence),
    RMessage = as.character(fit$message %||% ""),
    RLogLikQ31 = -as.numeric(fit$value),
    RQ31SolutionRLogLikQ61 = fit_q61$loglik,
    RGradientSupNormQ31 = max(abs(fit_gradient)),
    RSigma = expanded_r$sigma,
    RFunctionEvaluations = as.integer(fit$counts[["function"]]),
    RGradientEvaluations = as.integer(fit$counts[["gradient"]]),
    RQ61ConvergenceCode = as.integer(fit61$convergence),
    RQ61Message = as.character(fit61$message %||% ""),
    ROptimizedLogLikQ61 = -as.numeric(fit61$value),
    RGradientSupNormQ61 = max(abs(fit61_gradient)),
    RQ61Sigma = expanded_r61$sigma,
    RQ61FunctionEvaluations = as.integer(fit61$counts[["function"]]),
    RQ61GradientEvaluations = as.integer(fit61$counts[["gradient"]]),
    stringsAsFactors = FALSE
  )

  for (coordinate_index in seq_along(coordinate_names)) {
    gradient_index <- gradient_index + 1L
    gradient_rows[[gradient_index]] <- data.frame(
      RunId = run_id,
      Coordinate = coordinate_names[[coordinate_index]],
      PythonSolutionRGradientQ31 = python_gradient[[coordinate_index]],
      RQ31SolutionRGradientQ31 = fit_gradient[[coordinate_index]],
      RQ61SolutionRGradientQ61 = fit61_gradient[[coordinate_index]],
      stringsAsFactors = FALSE
    )
  }

  blocks <- list(
    Rater = c("R01", "R02", "R03", "R04"),
    Task = c("T01", "T02", "T03"),
    Criterion = c("C01", "C02")
  )
  for (block in names(blocks)) {
    levels <- blocks[[block]]
    for (level_index in seq_along(levels)) {
      parameter_index <- parameter_index + 1L
      parameter_rows[[parameter_index]] <- data.frame(
        RunId = run_id, Block = block, Level = levels[[level_index]],
        PythonEstimate = expanded_python[[block]][[level_index]],
        REstimateQ31 = expanded_r[[block]][[level_index]],
        REstimateQ61 = expanded_r61[[block]][[level_index]], stringsAsFactors = FALSE
      )
    }
  }
  for (criterion_index in seq_len(2L)) {
    for (step_index in seq_len(3L)) {
      parameter_index <- parameter_index + 1L
      parameter_rows[[parameter_index]] <- data.frame(
        RunId = run_id, Block = "Step",
        Level = paste0(c("C01", "C02")[[criterion_index]], "::", step_index),
        PythonEstimate = expanded_python$steps[criterion_index, step_index],
        REstimateQ31 = expanded_r$steps[criterion_index, step_index],
        REstimateQ61 = expanded_r61$steps[criterion_index, step_index], stringsAsFactors = FALSE
      )
    }
  }
}

runs <- do.call(rbind, run_rows)
parameters <- do.call(rbind, parameter_rows)
gradients <- do.call(rbind, gradient_rows)
parameters$DifferenceRQ31MinusPython <- parameters$REstimateQ31 - parameters$PythonEstimate
parameters$DifferenceRQ61MinusRQ31 <- parameters$REstimateQ61 - parameters$REstimateQ31
utils::write.csv(runs, file.path(output_dir, "crossfit_runs.csv"), row.names = FALSE)
utils::write.csv(parameters, file.path(output_dir, "crossfit_parameters.csv"), row.names = FALSE)
utils::write.csv(gradients, file.path(output_dir, "crossfit_gradients.csv"), row.names = FALSE)
software_versions <- extSoftVersion()
software_value <- function(name) {
  if (name %in% names(software_versions)) software_versions[[name]] else NA_character_
}
utils::write.csv(
  data.frame(
    RVersion = as.character(getRversion()),
    Platform = R.version$platform,
    BLAS = software_value("BLAS"),
    LAPACK = software_value("LAPACK"),
    SelectedVectors = paste(vector_ids, collapse = ","),
    Datasets = nrow(runs),
    stringsAsFactors = FALSE
  ),
  file.path(output_dir, "runtime_identity.csv"),
  row.names = FALSE
)

File: validation/known_assignment_mml_crossfit.py
#!/usr/bin/env python3
"""Cross-evaluate the independent base-R free-SD PCM MML refits."""

from __future__ import annotations

import argparse
import hashlib
import json
from pathlib import Path
import sys
from typing import Any, Iterable

import numpy as np
from numpy.polynomial.hermite import hermgauss
import pandas as pd
from scipy.special import logsumexp


ROOT = Path(__file__).resolve().parents[1]
if str(ROOT) not in sys.path:
    sys.path.insert(0, str(ROOT))

from validation.operating_characteristics_facets import sha256_file  # noqa: E402


SCHEMA_VERSION = "known_assignment_mml_crossfit_v1"
R_SCRIPT = ROOT / "validation" / "known_assignment_mml_crossfit.R"
PARAMETER_BLOCK_LEVELS = {
    "Rater": ("R01", "R02", "R03", "R04"),
    "Task": ("T01", "T02", "T03"),
    "Criterion": ("C01", "C02"),
    "Step": ("C01::1", "C01::2", "C01::3", "C02::1", "C02::2", "C02::3"),
}


def _json_dump(path: Path, value: Any) -> None:
    path.write_text(
        json.dumps(value, ensure_ascii=False, indent=2, sort_keys=True) + "\n",
        encoding="utf-8",
    )


def quadrature(points: int, sigma: float) -> tuple[np.ndarray, np.ndarray]:
    nodes, weights = hermgauss(int(points))
    return np.sqrt(2.0) * nodes * float(sigma), weights / np.sqrt(np.pi)


def parameter_surface(frame: pd.DataFrame, column: str) -> dict[str, Any]:
    output: dict[str, Any] = {}
    for block, levels in PARAMETER_BLOCK_LEVELS.items():
        selected = frame.loc[frame["Block"].astype(str).eq(block)].set_index("Level")
        if set(selected.index.astype(str)) != set(levels):
            raise ValueError(f"Cross-fit parameter levels changed for {block}")
        values = pd.to_numeric(selected[column], errors="raise")
        if block == "Step":
            output[block] = {
                criterion: np.array(
                    [float(values.loc[f"{criterion}::{step}"]) for step in (1, 2, 3)],
                    dtype=float,
                )
                for criterion in ("C01", "C02")
            }
        else:
            output[block] = {
                level: float(values.loc[level]) for level in levels
            }
    return output


def marginal_loglik(
    ratings: pd.DataFrame,
    surface: dict[str, Any],
    *,
    sigma: float,
    points: int,
) -> float:
    """Protocol-level Python evaluator independent of the app optimizer."""

    if not np.isfinite(sigma) or sigma <= 0:
        raise ValueError("Population SD must be positive and finite")
    data = ratings.copy()
    score = pd.to_numeric(data["Score"], errors="raise").astype(int).to_numpy()
    if not np.isin(score, (0, 1, 2, 3)).all():
        raise ValueError("Cross-fit score map must remain 0..3")
    base = -(
        data["Rater"].astype(str).map(surface["Rater"]).to_numpy(dtype=float)
        + data["Task"].astype(str).map(surface["Task"]).to_numpy(dtype=float)
        + data["Criterion"].astype(str).map(surface["Criterion"]).to_numpy(dtype=float)
    )
    if not np.isfinite(base).all():
        raise ValueError("Cross-fit facet map contains missing levels")
    criteria = data["Criterion"].astype(str).to_numpy()
    cumulative = {
        criterion: np.concatenate([[0.0], np.cumsum(surface["Step"][criterion])])
        for criterion in ("C01", "C02")
    }
    thresholds = np.vstack([cumulative[criterion] for criterion in criteria])
    person_codes, people = pd.factorize(data["Person"].astype(str), sort=True)
    nodes, weights = quadrature(points, sigma)
    person_node = np.zeros((len(people), points), dtype=float)
    categories = np.arange(4, dtype=float)
    row_index = np.arange(len(data))
    for node_index, node in enumerate(nodes):
        logits = np.outer(node + base, categories) - thresholds
        row_ll = logits[row_index, score] - logsumexp(logits, axis=1)
        person_node[:, node_index] = np.bincount(
            person_codes, weights=row_ll, minlength=len(people)
        )
    return float(np.sum(logsumexp(np.log(weights)[None, :] + person_node, axis=1)))


def _maximum_absolute(frame: pd.DataFrame, column: str) -> float:
    return float(pd.to_numeric(frame[column], errors="coerce").abs().max())


def assess_crossfit(
    *,
    study_dir: Path,
    output_dir: Path,
    expected_datasets: int,
    tolerance: dict[str, float],
) -> dict[str, Any]:
    study_dir = study_dir.resolve()
    output_dir = output_dir.resolve()
    runs = pd.read_csv(output_dir / "crossfit_runs.csv")
    parameters = pd.read_csv(output_dir / "crossfit_parameters.csv")
    gradients = pd.read_csv(output_dir / "crossfit_gradients.csv")
    runtime = pd.read_csv(output_dir / "runtime_identity.csv")
    ratings_all = pd.read_csv(study_dir / "retained_input" / "generated_ratings.csv")
    if len(runs) != expected_datasets or runs["RunId"].nunique() != expected_datasets:
        raise ValueError("R cross-fit dataset denominator changed")
    if len(parameters) != expected_datasets * 15:
        raise ValueError("R cross-fit expanded-parameter denominator changed")
    if len(gradients) != expected_datasets * 12:
        raise ValueError("R cross-fit gradient denominator changed")

    evaluation_rows: list[dict[str, Any]] = []
    for run in runs.itertuples(index=False):
        run_id = str(run.RunId)
        parameter = parameters.loc[parameters["RunId"].astype(str).eq(run_id)]
        ratings = ratings_all.loc[
            ratings_all["RunId"].astype(str).eq(run_id),
            ["Person", "Rater", "Task", "Criterion", "Score"],
        ]
        python_surface = parameter_surface(parameter, "PythonEstimate")
        r31_surface = parameter_surface(parameter, "REstimateQ31")
        r61_surface = parameter_surface(parameter, "REstimateQ61")
        values = {
            "PythonSolutionPythonLogLikQ31": marginal_loglik(
                ratings, python_surface, sigma=float(run.PythonSigma), points=31
            ),
            "PythonSolutionPythonLogLikQ61": marginal_loglik(
                ratings, python_surface, sigma=float(run.PythonSigma), points=61
            ),
            "RQ31SolutionPythonLogLikQ31": marginal_loglik(
                ratings, r31_surface, sigma=float(run.RSigma), points=31
            ),
            "RQ31SolutionPythonLogLikQ61": marginal_loglik(
                ratings, r31_surface, sigma=float(run.RSigma), points=61
            ),
            "RQ61SolutionPythonLogLikQ61": marginal_loglik(
                ratings, r61_surface, sigma=float(run.RQ61Sigma), points=61
            ),
        }
        evaluation_rows.append({"RunId": run_id, **values})
    evaluations = pd.DataFrame(evaluation_rows).merge(
        runs, on="RunId", how="left", validate="one_to_one"
    )
    evaluations["AbsRAtPythonVsRecordedQ31"] = (
        evaluations["PythonSolutionRLogLikQ31"] - evaluations["PythonRecordedLogLik"]
    ).abs()
    evaluations["AbsPythonAtPythonVsRecordedQ31"] = (
        evaluations["PythonSolutionPythonLogLikQ31"] - evaluations["PythonRecordedLogLik"]
    ).abs()
    evaluations["AbsCrossLanguageAtPythonQ31"] = (
        evaluations["PythonSolutionPythonLogLikQ31"]
        - evaluations["PythonSolutionRLogLikQ31"]
    ).abs()
    evaluations["AbsCrossLanguageAtRQ31"] = (
        evaluations["RQ31SolutionPythonLogLikQ31"] - evaluations["RLogLikQ31"]
    ).abs()
    evaluations["AbsCrossLanguageAtRQ61"] = (
        evaluations["RQ61SolutionPythonLogLikQ61"]
        - evaluations["ROptimizedLogLikQ61"]
    ).abs()
    evaluations["RQ31ImprovementOverPythonQ31"] = (
        evaluations["RLogLikQ31"] - evaluations["PythonRecordedLogLik"]
    )
    evaluations["PythonQ61MinusQ31"] = (
        evaluations["PythonSolutionPythonLogLikQ61"]
        - evaluations["PythonSolutionPythonLogLikQ31"]
    )
    evaluations["RQ61OptimizedMinusRQ31EvaluatedQ61"] = (
        evaluations["ROptimizedLogLikQ61"]
        - evaluations["RQ31SolutionPythonLogLikQ61"]
    )

    parameter_q31 = _maximum_absolute(parameters, "DifferenceRQ31MinusPython")
    parameter_q61 = _maximum_absolute(parameters, "DifferenceRQ61MinusRQ31")
    q31_sigma = float((evaluations["RSigma"] - evaluations["PythonSigma"]).abs().max())
    q61_sigma = float((evaluations["RQ61Sigma"] - evaluations["RSigma"]).abs().max())
    gates = {
        "r_version_4_5_1": str(runtime.loc[0, "RVersion"]) == "4.5.1",
        "datasets_complete": len(evaluations) == expected_datasets,
        "r_q31_convergence_all": bool(evaluations["RConvergenceCode"].eq(0).all()),
        "r_q61_convergence_all": bool(evaluations["RQ61ConvergenceCode"].eq(0).all()),
        "r_at_python_matches_recorded_q31": float(
            evaluations["AbsRAtPythonVsRecordedQ31"].max()
        )
        <= tolerance["cross_language_loglik"],
        "python_at_python_matches_recorded_q31": float(
            evaluations["AbsPythonAtPythonVsRecordedQ31"].max()
        )
        <= tolerance["cross_language_loglik"],
        "cross_language_python_solution_q31": float(
            evaluations["AbsCrossLanguageAtPythonQ31"].max()
        )
        <= tolerance["cross_language_loglik"],
        "cross_language_r_solution_q31": float(
            evaluations["AbsCrossLanguageAtRQ31"].max()
        )
        <= tolerance["cross_language_loglik"],
        "cross_language_r_solution_q61": float(
            evaluations["AbsCrossLanguageAtRQ61"].max()
        )
        <= tolerance["cross_language_loglik"],
        "r_q31_terminal_gradient": float(evaluations["RGradientSupNormQ31"].max())
        <= tolerance["r_terminal_gradient_q31"],
        "r_q61_terminal_gradient": float(evaluations["RGradientSupNormQ61"].max())
        <= tolerance["r_terminal_gradient_q61"],
        "python_sigma_fixed_point": float(
            evaluations["PythonSigmaFixedPointResidual"].abs().max()
        )
        <= tolerance["python_sigma_fixed_point"],
        "r_q31_vs_python_parameter": parameter_q31 <= tolerance["r_q31_parameter"],
        "r_q31_vs_python_sigma": q31_sigma <= tolerance["r_q31_sigma"],
        "q61_vs_q31_parameter": parameter_q61 <= tolerance["q61_parameter"],
        "q61_vs_q31_sigma": q61_sigma <= tolerance["q61_sigma"],
        "r_q31_loglik_improvement_bounded": float(
            evaluations["RQ31ImprovementOverPythonQ31"].max()
        )
        <= tolerance["r_q31_loglik_improvement"],
    }
    metrics = {
        "maximum_abs_r_at_python_vs_recorded_q31": float(
            evaluations["AbsRAtPythonVsRecordedQ31"].max()
        ),
        "maximum_abs_python_at_python_vs_recorded_q31": float(
            evaluations["AbsPythonAtPythonVsRecordedQ31"].max()
        ),
        "maximum_cross_language_loglik_difference": float(
            evaluations[
                [
                    "AbsCrossLanguageAtPythonQ31",
                    "AbsCrossLanguageAtRQ31",
                    "AbsCrossLanguageAtRQ61",
                ]
            ].to_numpy(dtype=float).max()
        ),
        "maximum_r_q31_gradient_sup_norm": float(evaluations["RGradientSupNormQ31"].max()),
        "maximum_r_q61_gradient_sup_norm": float(evaluations["RGradientSupNormQ61"].max()),
        "maximum_abs_python_sigma_fixed_point_residual": float(
            evaluations["PythonSigmaFixedPointResidual"].abs().max()
        ),
        "maximum_abs_r_q31_minus_python_parameter": parameter_q31,
        "maximum_abs_r_q31_minus_python_sigma": q31_sigma,
        "maximum_abs_r_q61_minus_q31_parameter": parameter_q61,
        "maximum_abs_r_q61_minus_q31_sigma": q61_sigma,
        "maximum_r_q31_loglik_improvement_over_python": float(
            evaluations["RQ31ImprovementOverPythonQ31"].max()
        ),
        "maximum_abs_python_q61_minus_q31_loglik": float(
            evaluations["PythonQ61MinusQ31"].abs().max()
        ),
    }
    evaluations.to_csv(output_dir / "crossfit_cross_evaluations.csv", index=False, lineterminator="\n")
    identity = {
        "schema_version": f"{SCHEMA_VERSION}_identity_v1",
        "study_dir": str(study_dir),
        "expected_datasets": expected_datasets,
        "r_script_sha256": sha256_file(R_SCRIPT),
        "python_verifier_sha256": sha256_file(Path(__file__).resolve()),
        "source_sha256": {
            filename: sha256_file(output_dir / filename)
            for filename in (
                "crossfit_runs.csv",
                "crossfit_parameters.csv",
                "crossfit_gradients.csv",
                "runtime_identity.csv",
            )
        },
        "retained_input_sha256": {
            filename: sha256_file(study_dir / "retained_input" / filename)
            for filename in ("manifest.csv", "generated_ratings.csv", "attempt_manifest.csv")
        },
    }
    _json_dump(output_dir / "verification_identity.json", identity)
    assessment = {
        "schema_version": SCHEMA_VERSION,
        "pass": bool(all(gates.values())),
        "gates": {key: bool(value) for key, value in gates.items()},
        "tolerance": tolerance,
        "metrics": metrics,
        "datasets": expected_datasets,
        "r_version": str(runtime.loc[0, "RVersion"]),
        "r_script_sha256": identity["r_script_sha256"],
        "python_verifier_sha256": identity["python_verifier_sha256"],
        "claim_boundary": "Same-estimand numerical cross-fit; not FACETS parity, endpoint confirmation, DGP replication, or estimator ranking.",
    }
    _json_dump(output_dir / "assessment.json", assessment)
    return assessment


def default_tolerance() -> dict[str, float]:
    """Candidate tolerances; freeze them in registration after fixture qualification."""

    return {
        "cross_language_loglik": 1e-9,
        "r_terminal_gradient_q31": 1e-3,
        "r_terminal_gradient_q61": 1e-2,
        "python_sigma_fixed_point": 1e-3,
        "r_q31_parameter": 0.01,
        "r_q31_sigma": 0.01,
        "q61_parameter": 0.01,
        "q61_sigma": 0.01,
        "r_q31_loglik_improvement": 0.01,
    }


def parse_args(argv: Iterable[str] | None = None) -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--study", type=Path, required=True)
    parser.add_argument("--output", type=Path, required=True)
    parser.add_argument("--expected-datasets", type=int, required=True)
    return parser.parse_args(argv)


def main() -> None:
    args = parse_args()
    result = assess_crossfit(
        study_dir=args.study,
        output_dir=args.output,
        expected_datasets=args.expected_datasets,
        tolerance=default_tolerance(),
    )
    print(json.dumps(result, ensure_ascii=False, indent=2, sort_keys=True))


if __name__ == "__main__":
    main()
